client/src/main/java/io/pravega/client/connection/impl/RawClient.java

```diff
@@ -98,23 +98,25 @@ public void authTokenCheckFailed(WireCommands.AuthTokenCheckFailed authTokenChec
 
     public RawClient(PravegaNodeUri uri, ConnectionPool connectionPool) {
         this.segmentId = null;
-        this.connection = connectionPool.getClientConnection(flow, uri, responseProcessor)
-        .exceptionally(e -> {
+        this.connection = new CompletableFuture<>();
+        connectionPool.getClientConnection(flow, uri, responseProcessor, this.connection);
+        this.connection.exceptionally(e -> {
             log.warn("Exception observed while attempting to obtain a connection to segment store {}", uri, e);
             throw new CompletionException(new ConnectionFailedException(e));
         });
-        Futures.exceptionListener(connection, e -> closeConnection(e));
+        Futures.exceptionListener(connection, this::closeConnection);
```

Comment:
Shouldn't this and the `exceptionally` be moved above the `getClientConnection` call now that that is possible?